### ISSUE TITLE
problem with thetapv2 and wgsmax (uncorrectly filtered?)

### ISSUE THREAD

**Issue by oloapinivad** [goord/ece2cmor3#146]
Hi again,

sorry to insist in my debugging (now in PRIMAVERA tables). 
I noticed that two variables are encountering issues due to wrong levels when filtered. 

`````
ERROR:ece2cmor3.grib_filter:Field missing in the first day of file: code 3.128, level type 1, level 0. Dismissing task thetapv2 in table Prim6hrPt
ERROR:ece2cmor3.grib_filter:Field missing in the first day of file: code 49.128, level type 105, level 10. Dismissing task wsgmax in table Prim6hr
`````
However, when I check the original files, I get:

`````
grib_ls -w,paramId=3/49 -p,paramId,levelType,level,dataDate,dataTime ICMGGqctr+194912 
ICMGGqctr+194912
paramId     levelType   level       dataDate    dataTime    
49          sfc         0           19500101    0          
3           pv          2000        19500101    0          
2 of 129 grib messages in ICMGGqctr+194912

`````

Apparently both thetapv and wgsmax are in the file, on level pv (i.e. level 117) and on level sfc (i.e. level 1, if I am decoding grib correctly). 

I speculate this is related on the way they are filtered. However, while potential temperature on PV surface it is usually defined on 2PVU units (so that at the surface is wrong), I am not sure that wgsmax problem is the same. 
Var 49 it is saved at the surface on our simulation, while on the table is at _height10m_. However, since also uas and vas are defined at the surface in the model and then they are converted to _height10m_, it may be that there is a mismatch in the level type too.
Thanks for any help
Best,
Paolo

**Comment by goord** [comment]
Yes you diagnosed the problem correctly: during filtering there is 'matching' procedure to decide which grib messages are relevant for which cmor variable. This of course involves looking at the level type and seeing whether a match can be made. However, there many hard-coded boundary cases, such as 2m-temperature which IFS says is at the surface and CMOR says is at height2m. wgsmax is also such a case. PV levels are not included in the matching at all, I will try to fix this now.

**Comment by goord** [comment]
Hi Paolo, this should work after rev. b5cacb3b80441219d7d8f8078f80504d6b41676d

**Comment by oloapinivad** [comment]
Hi @goord,

thanks a lot. I have just tested it, and it works for thetapv smooth. 
Unfortunately, for wgsmax I think we have again the same issue similar to the one in #127
Actually, it is the same one I have with sfcWindmax.
I am reporting:

`````
ERROR:ece2cmor3.postproc:Unsupported combination of frequency 6hr with time operators [u'maximum'] encountered
ERROR:ece2cmor3.postproc:Unsupported combination of frequency 6hr with time operators [u'maximum'] encountered
INFO:ece2cmor3.postproc:Post-processing target sfcWindmax in table Prim6hr from file /marconi_scratch/userexternal/pdavini0/tmp_cmor/qctr_29540/tmp_1950_1/cmorized/PRIMAVERA/165.128.105_166.128.105.3 with cdo command -expr,'var214=sqrt(sqr(var165)+sqr(var166))' -setgridtype,regular -selcode,165,166
ERROR:ece2cmor3.cmor_task:Attempt to increase failed status for task sfcWindmax in Prim6hr ignored
INFO:ece2cmor3.postproc:Post-processing target wsgmax in table Prim6hr from file /marconi_scratch/userexternal/pdavini0/tmp_cmor/qctr_29540/tmp_1950_1/cmorized/PRIMAVERA/49.128.105.3 with cdo command -setgridtype,regular -selcode,49
ERROR:ece2cmor3.cmor_task:Attempt to increase failed status for task wsgmax in Prim6hr ignored

`````

Should I fix manually the table as you suggested in the #127 ? 
Best,
Paolo

**Comment by goord** [comment]
Well the problem here is physical: the 6-hour maximal theta is requested, but only 6-hourly snapshots have been saved, so IMO it is not possible to produce this. The sfcWindmax could be regarded as the wind gust right (grib code 49)?

**Comment by oloapinivad** [comment]
I see your point. Investigating a bit I am finding here that https://software.ecmwf.int/wiki/display/OIFS/How+to+control+OpenIFS+output code 49 is defined as:
 
````
Maximum 3 second wind at 10 m height as defined by WMO. 
Parametrization represents turbulence only before 01/10/2008; thereafter effects of convection are included.
The 3 s gust is computed every time step and the maximum is kept since the last postprocessing.
````

So, as far as I understand 6hr output should include already the maximum evaluated on the previous 6hr (i.e. the last postprocessing). Actually this is the same as tasmin and tasmax. 

On the contrary, I think you are right with sfcWindmax, which is actually defined as the max of the module of 10m wind. Unfortunately the machine is now offline so I cannot check if we have this at 3hr frequency or if I am trying without success to store it at lower frequency (i.e. daily). More will follow as soon as it gets back online.

**Comment by goord** [comment]
Ok, I will change the parameter table to use 49 as sfcwindmax.

**Comment by oloapinivad** [comment]
Hi @goord,

I further investigated and according to what I found the final response is no, you should not use 49 for sfcwindmax. 

Actually wind speed and wind gust have two WMO different definitions (see here http://mmisw.org/ont/cf/parameter/wind_speed_of_gust). Code 49 is evaluating with *parametrization* each 3s effect of the gustiness, so it is not the maximum of the wind speed at each time step. Conversely, there is no maximum of the wind computed stored each time step in IFS.
This means that the whole current definition for sfcWindmax in the end is right, and what I get is an issue of my outputs (as you said in the beginning, too low frequency). 
However, I feel that there is something wrong for wsgmax, since code 49 is already a maximum, and if we save it every 6hrs this would be already ok in the IFS code.

I hope I was clear enough, let me know if I can help you further.

**Comment by oloapinivad** [comment]
Sorry if I insist, 

but I have access again to the machine. I have noticed that in my output I have data at 3hr frequency for code 49, 165 and 166 (10fg, u10 and v10). So it should be possible to compute the maximum on 6hr, so I speculate that perhaps something is wrong in the filtering also for those vars?

`````
[pdavini0@r000u06l01 linkdata]$ grib_ls -w,paramId=49/165/166 -p,paramId,levelType,level,dataDate,dataTime ICMGGqctr+195001
ICMGGqctr+195001
paramId     levelType   level       dataDate    dataTime    
49          sfc         0           19500101    300        
165         sfc         0           19500101    300        
166         sfc         0           19500101    300        
49          sfc         0           19500101    600        
165         sfc         0           19500101    600        
166         sfc         0           19500101    600        
49          sfc         0           19500101    900        
165         sfc         0           19500101    900        
166         sfc         0           19500101    900        
49          sfc         0           19500101    1200       
165         sfc         0           19500101    1200       
166         sfc         0           19500101    1200       
49          sfc         0           19500101    1500       
165         sfc         0           19500101    1500       
166         sfc         0           19500101    1500       
49          sfc         0           19500101    1800       
165         sfc         0           19500101    1800       
166         sfc         0           19500101    1800       
49          sfc         0           19500101    2100       
165         sfc         0           19500101    2100       
166         sfc         0           19500101    2100       
``````

**Comment by goord** [comment]
Hi Paolo, no I just don't support it because I think it doesn't make much sense to compute 6hr maximum values from sampling of 3-hourly fields. Technically we could do it, but will the result be anywhere near the true maximum (especially for the wind)?

**Comment by oloapinivad** [comment]
You are right, for wind speed (sfcWindmax) it's pretty non-sense.

Conversly, for the wind gust (wsgmax) I think this should be supported, since - using 3hr output frequency - this is already the maximum over the previously 3hr. 
So in this case we will be catching _exactly_ the maximum of wind gust doing the maximum between time 0300 and 0600.

**Comment by goord** [comment]
Agreed, I will add this to the code
